domain/macaroon/service/rootkeys.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/bakery"
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/bakery/dbrootkeystore"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/domain/macaroon"
+	macaroonerrors "github.com/juju/juju/domain/macaroon/errors"
+)
+
+// RootKeyState describes the persistence layer for
+// macaroon root keys
+type RootKeyState interface {
+	// GetKey gets the key with a given id from state. If not key is found, a
+	// macaroonerrors.KeyNotFound error is returned.
+	GetKey(ctx context.Context, id []byte) (macaroon.RootKey, error)
+
+	// FindLatestKey returns the most recently created root key k following all
+	// the conditions:
+	//
+	// k.Created >= createdAfter
+	// k.Expires >= expiresAfter
+	// k.Expires <= expiresBefore
+	//
+	// If no such key was found, return a macaroonerrors.KeyNotFound error
+	FindLatestKey(ctx context.Context, createdAfter, expiresAfter, expiresBefore time.Time) (macaroon.RootKey, error)
+
+	// InsertKey inserts the given root key into state. If a key with matching
+	// id already exists, return a macaroonerrors.KeyAlreadyExists error.
+	InsertKey(ctx context.Context, key macaroon.RootKey) error
+}
+
+// RootKeyService provides the API for macaroon root key storage
+//
+// RootKeyService satisfies dbrootkeystore.Backing and dbrootkeystore.ContextBacking
+// https://github.com/go-macaroon-bakery/macaroon-bakery/blob/f9b21e15a2ed91756aa172972c7178992c7fe6d1/bakery/dbrootkeystore/rootkey.go#L48-L95
+//
+// This means RootKeyService can be used to construct a bakery.RootKeyStore.
+//
+// NOTE: We implement dbrootkeystore.Backing with stub methods. This is because
+// RootKeyStore only requires a ContextBacking, but due to pecularities with
+// the RootKeyStore constructor, we also need to implement Backing.
+// TODO(jack-w-shaw): Once https://github.com/go-macaroon-bakery/macaroon-bakery/pull/301
```

Comment:
That appears to be merged.

Reply:
It was this morning yes. I will update the commen